spec/controllers/queue_items_controller_spec.rb

```diff
@@ -0,0 +1,203 @@
+        get :index
+        expect(assigns(:queue_items)).to match_array([q1, q2])
+      end
+    end
+      
+    context "When no user is signed in" do
+      it "redirects to welcome_path" do
+        get :index
+        expect(response).to redirect_to welcome_path
+      end
+    end
+  end
+
+  describe "POST create" do
+    let(:video) { Fabricate(:video) }
+
+    context "when a user is signed in" do
+      before do
+        session[:current_user_id] = pete.id
+        request.env['HTTP_REFERER'] = "from_whence_I_came"
+      end
+      
+      context "when the video is not yet in the user's queue" do
+        before do
+          session[:current_user_id] = pete.id
+          request.env['HTTP_REFERER'] = "from_whence_I_came"
+        end
+          
+        it "creates an instance of QueueItem" do
+          post :create, video_id: video.id
+          expect(QueueItem.count).to eq(1)
+        end
+
+        it "associates new queue item with current user" do
+          post :create, video_id: video.id
+          expect(QueueItem.first.user).to eq(pete)
+        end
+
+        it "associates new queue item with video" do
+          post :create, video_id: video.id
+          expect(QueueItem.first.video).to eq(video)
+        end
+
+        it "assigns the new QueueItem one greater than the highest position of existing queue items" do
+          QueueItem.create(position: 1, user: pete, video: Fabricate(:video))
+          post :create, video_id: video.id
+          expect(QueueItem.last.position).to eq(2)
+        end
+
+        it "sets flash[:success]" do
+          post :create, video_id: video.id
+          expect(flash[:success]).to be_present
+        end
+
+        it "redirects back to the previous page" do
+          post :create, video_id: video.id
+          expect(response).to redirect_to "from_whence_I_came"
+        end
+      end
+
+      context "when the video is already in the user's queue" do
+        before do 
+          QueueItem.create(position: 1, user: pete, video: video)
+          post :create, video_id: video.id
+        end
+
+        it "doesn't save the queue item" do
+          expect(QueueItem.count).to eq(1)
+        end
+
+        it "sets flash[:danger] message" do
+          expect(flash[:danger]).to be_present
+        end
+
+        it "redirects back to the previous page" do
+          expect(response).to redirect_to "from_whence_I_came"
+        end
+      end
+    end
+
+    context "when no user is signed in" do
+      it "redirects_to welcome_path" do
+        post :create, video_id: video.id
+        expect(response).to redirect_to(welcome_path)
+      end
+    end
+  end
+
+  describe "DELETE destroy" do
+    let(:pete) { Fabricate(:user) }
+    let(:james) { Fabricate(:user) }
+    let (:video1) { Fabricate(:video) }
+    let (:video2) { Fabricate(:video) }
+    let!(:q1) { QueueItem.create(position: 1, user: pete, video: video1) }
```

Comment:
Being lazy.  Since everything needs to be set except the position anyway, I just went ahead and did it by hand.  I also have uniqueness validation on `position`, so I have to be careful about not having overlapping position numbers, too.  I can fix it, though.